testutil/server.go

```diff
@@ -0,0 +1,434 @@
+	Address string   `json:",omitempty"`
+	Port    int      `json:",omitempty"`
+}
+
+// TestCheck is used to serialize a check definition.
+type TestCheck struct {
+	ID        string `json:",omitempty"`
+	Name      string `json:",omitempty"`
+	ServiceID string `json:",omitempty"`
+	TTL       string `json:",omitempty"`
+}
+
+// TestKVResponse is what we use to decode KV data.
+type TestKVResponse struct {
+	Value string
+}
+
+// TestServer is the main server wrapper struct.
+type TestServer struct {
+	PID    int
+	Config *TestServerConfig
+	t      *testing.T
+
+	HTTPAddr string
+	LANAddr  string
+	WANAddr  string
+
+	HttpClient *http.Client
+}
+
+// NewTestServer is an easy helper method to create a new Consul
+// test server with the most basic configuration.
+func NewTestServer(t *testing.T) *TestServer {
+	return NewTestServerConfig(t, nil)
+}
+
+// NewTestServerConfig creates a new TestServer, and makes a call to
+// an optional callback function to modify the configuration.
+func NewTestServerConfig(t *testing.T, cb ServerConfigCallback) *TestServer {
+	if path, err := exec.LookPath("consul"); err != nil || path == "" {
+		t.Skip("consul not found on $PATH, skipping")
+	}
+
+	dataDir, err := ioutil.TempDir("", "consul")
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	configFile, err := ioutil.TempFile("", "consul")
+	if err != nil {
+		defer os.RemoveAll(dataDir)
+		t.Fatalf("err: %s", err)
+	}
+	defer os.Remove(configFile.Name())
+
+	consulConfig := defaultServerConfig()
+	consulConfig.DataDir = dataDir
+
+	if cb != nil {
+		cb(consulConfig)
+	}
+
+	configContent, err := json.Marshal(consulConfig)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	if _, err := configFile.Write(configContent); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	configFile.Close()
+
+	// Start the server
+	cmd := exec.Command("consul", "agent", "-config-file", configFile.Name())
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	var httpAddr string
+	var client *http.Client
+	if strings.HasPrefix(consulConfig.Addresses.HTTP, "unix://") {
+		httpAddr = consulConfig.Addresses.HTTP
+		client = &http.Client{
+			Transport: &http.Transport{
+				Dial: func(_, _ string) (net.Conn, error) {
+					return net.Dial("unix", httpAddr[7:])
+				},
+			},
+		}
+	} else {
+		httpAddr = fmt.Sprintf("127.0.0.1:%d", consulConfig.Ports.HTTP)
+		client = http.DefaultClient
+	}
+
+	server := &TestServer{
+		Config: consulConfig,
+		PID:    cmd.Process.Pid,
+		t:      t,
+
+		HTTPAddr: httpAddr,
+		LANAddr:  fmt.Sprintf("127.0.0.1:%d", consulConfig.Ports.SerfLan),
+		WANAddr:  fmt.Sprintf("127.0.0.1:%d", consulConfig.Ports.SerfWan),
+
+		HttpClient: client,
+	}
+
+	// Wait for the server to be ready
+	server.waitForLeader()
+
+	return server
+}
+
+// Stop stops the test Consul server, and removes the Consul data
+// directory once we are done.
+func (s *TestServer) Stop() {
```

Comment:
I think the reason I did this was to stop dead in our tracks if we for some reason couldn't kill a Consul instance to prevent spinning up like 80 instances and then all of them just sitting in the background. It's probably not a common case. We could also save the `t` from `NewTestServer` and use that as well.